File: go/internal/parser/insert.go
package parser

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/vbauerster/mpb/v8"
	"github.com/vbauerster/mpb/v8/decor"
	"github.com/younes/sqldumpdiff/internal/logger"
)

// InsertRow represents a parsed INSERT row with table, columns, and data map.
type InsertRow struct {
	Table   string
	Columns []string
	Data    map[string]string
}

// InsertParser handles parsing of INSERT statements from dumps.
type InsertParser struct {
}

// NewInsertParser creates a new insert parser
func NewInsertParser() *InsertParser {
	return &InsertParser{}
}

// InsertAccumulator helps build complete INSERT statements across multiple lines.
// It tracks quote/paren state to know when a statement is complete.
type InsertAccumulator struct {
	buffer              strings.Builder
	inSingleQuote       bool
	inDoubleQuote       bool
	escapeNext          bool
	parenDepth          int
	inInsert            bool
	statementsProcessed int
}

// ProcessLine processes a line and returns any complete INSERT statements found.
func (acc *InsertAccumulator) ProcessLine(line string) []string {
	var results []string

	// Check if starting a new INSERT
	if !acc.inInsert && strings.Contains(strings.ToUpper(line), "INSERT INTO") {
		acc.buffer.Reset()
		acc.buffer.WriteString(line)
		acc.buffer.WriteString("\n")
		acc.inInsert = true
		acc.inSingleQuote = false
		acc.inDoubleQuote = false
		acc.escapeNext = false
		acc.parenDepth = 0

		acc.processLineContent(line)

		if strings.HasSuffix(strings.TrimSpace(line), ";") && acc.parenDepth == 0 {
			results = append(results, acc.buffer.String())
			acc.buffer.Reset()
			acc.inInsert = false
			acc.statementsProcessed++
		}
		return results
	}

	// Continue processing if in INSERT
	if acc.inInsert {
		acc.buffer.WriteString(line)
		acc.buffer.WriteString("\n")
		acc.processLineContent(line)

		if strings.HasSuffix(strings.TrimSpace(line), ";") && acc.parenDepth == 0 {
			results = append(results, acc.buffer.String())
			acc.buffer.Reset()
			acc.inInsert = false
			acc.statementsProcessed++
		}
	}

	return results
}

// processLineContent tracks quote state and parenthesis depth while scanning a line.
func (acc *InsertAccumulator) processLineContent(line string) {
	for _, c := range line {
		if acc.escapeNext {
			acc.escapeNext = false
			continue
		}
		if c == '\\' {
			acc.escapeNext = true
			continue
		}
		if c == '\'' && !acc.inDoubleQuote {
			acc.inSingleQuote = !acc.inSingleQuote
		} else if c == '"' && !acc.inSingleQuote {
			acc.inDoubleQuote = !acc.inDoubleQuote
		} else if !acc.inSingleQuote && !acc.inDoubleQuote {
			if c == '(' {
				acc.parenDepth++
			} else if c == ')' {
				acc.parenDepth--
			}
		}
	}
}

// Finalize returns any incomplete statement at EOF.
func (acc *InsertAccumulator) Finalize() []string {
	var results []string
	if acc.inInsert && acc.buffer.Len() > 0 {
		results = append(results, acc.buffer.String())
		acc.statementsProcessed++
	}
	return results
}

// ParseInsertsStream reads INSERT statements from a file and calls onRow for each row.
// This is the streaming entrypoint used by the comparer for large files.
func (ip *InsertParser) ParseInsertsStream(filename string, columns map[string][]string, p *mpb.Progress, label string, onRow func(*InsertRow)) error {
	logger.Debug("ParseInsertsStream: Opening file %s", filename)
	file, err := os.Open(filename)
	if err != nil {
		logger.Error("ParseInsertsStream: Failed to open file %s: %v", filename, err)
		return err
	}
	defer file.Close()

	// Get file size for progress bar
	fi, err := file.Stat()
	if err != nil {
		logger.Error("ParseInsertsStream: Failed to stat file: %v", err)
		return err
	}
	fileSize := fi.Size()
	logger.Debug("ParseInsertsStream: File size: %d bytes", fileSize)

	var bar *mpb.Bar
	if p != nil {
		desc := label
		if desc == "" {
			desc = fmt.Sprintf("Parsing %s", filepath.Base(filename))
		}
		bar = p.New(
			fileSize,
			mpb.BarStyle().Lbound("[").Filler("█").Tip("█").Padding(" ").Rbound("]"),
			mpb.PrependDecorators(
				decor.Name(desc, decor.WC{W: 20, C: decor.DindentRight | decor.DextraSpace}),
				decor.CountersKibiByte("% .2f / % .2f", decor.WC{W: 18, C: decor.DindentRight | decor.DextraSpace}),
			),
			mpb.AppendDecorators(
				decor.Percentage(decor.WC{W: 5}),
			),
		)
	}

	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 0, 64*1024), 16*1024*1024) // 16MB buffer for large lines

	acc := &InsertAccumulator{}
	bytesRead := int64(0)
	for scanner.Scan() {
		line := scanner.Text()
		lineSize := int64(len(line)) + 1
		bytesRead += lineSize
		if bar != nil {
			bar.IncrBy(int(lineSize))
		}

		// Process the line and get any complete INSERT statements
		statements := acc.ProcessLine(line)
		for _, stmt := range statements {
			rows, err := ip.ExpandInsert(stmt, columns)
			if err != nil {
				logger.Debug("ParseInsertsStream: Failed to expand INSERT statement: %v", err)
				continue
			}

			for _, row := range rows {
				onRow(row)
			}
		}
	}

	// Handle any incomplete statement at EOF
	statements := acc.Finalize()
	for _, stmt := range statements {
		rows, err := ip.ExpandInsert(stmt, columns)
		if err != nil {
			logger.Debug("ParseInsertsStream: Failed to expand INSERT statement: %v", err)
			continue
		}

		for _, row := range rows {
			onRow(row)
		}
	}

	if err := scanner.Err(); err != nil {
		logger.Error("ParseInsertsStream: Scanner error: %v", err)
		return err
	}
	if bar != nil {
		bar.SetTotal(bytesRead, true)
	}

	logger.Debug("ParseInsertsStream: Processed %d INSERT statements", acc.statementsProcessed)
	return nil
}

// ParseInserts reads INSERT statements and returns an in-memory map of rows.
// This is used for smaller workflows; large-file flows use ParseInsertsStream.
func (ip *InsertParser) ParseInserts(filename string, columns map[string][]string, p *mpb.Progress) (map[string][]*InsertRow, error) {
	logger.Debug("ParseInserts: Opening file %s", filename)
	file, err := os.Open(filename)
	if err != nil {
		logger.Error("ParseInserts: Failed to open file %s: %v", filename, err)
		return nil, err
	}
	defer file.Close()

	// Get file size for progress bar
	fi, err := file.Stat()
	if err != nil {
		logger.Error("ParseInserts: Failed to stat file: %v", err)
		return nil, err
	}
	fileSize := fi.Size()
	logger.Debug("ParseInserts: File size: %d bytes", fileSize)

	var bar *mpb.Bar
	if p != nil {
		bar = p.New(
			fileSize,
			mpb.BarStyle().Lbound("[").Filler("█").Tip("█").Padding(" ").Rbound("]"),
			mpb.PrependDecorators(
				decor.Name(fmt.Sprintf("Parsing %s", filepath.Base(filename)), decor.WC{W: 20, C: decor.DindentRight | decor.DextraSpace}),
				decor.CountersKibiByte("% .2f / % .2f", decor.WC{W: 18, C: decor.DindentRight | decor.DextraSpace}),
			),
			mpb.AppendDecorators(
				decor.Percentage(decor.WC{W: 5}),
			),
		)
	}

	tableRows := make(map[string][]*InsertRow)
	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 0, 64*1024), 16*1024*1024) // 16MB buffer for large lines

	acc := &InsertAccumulator{}
	rowsExtracted := 0
	bytesRead := int64(0)
	for scanner.Scan() {
		line := scanner.Text()
		lineSize := int64(len(line)) + 1
		bytesRead += lineSize // +1 for newline
		if bar != nil {
			bar.IncrBy(int(lineSize))
		}

		// Process the line and get any complete INSERT statements
		statements := acc.ProcessLine(line)
		for _, stmt := range statements {
			rows, err := ip.ExpandInsert(stmt, columns)
			if err != nil {
				logger.Debug("ParseInserts: Failed to expand INSERT statement: %v", err)
				continue
			}

			for _, row := range rows {
				tableRows[row.Table] = append(tableRows[row.Table], row)
				rowsExtracted++
			}
		}
	}

	// Handle any incomplete statement at EOF
	statements := acc.Finalize()
	for _, stmt := range statements {
		rows, err := ip.ExpandInsert(stmt, columns)
		if err != nil {
			logger.Debug("ParseInserts: Failed to expand INSERT statement: %v", err)
			continue
		}

		for _, row := range rows {
			tableRows[row.Table] = append(tableRows[row.Table], row)
			rowsExtracted++
		}
	}

	if err := scanner.Err(); err != nil {
		logger.Error("ParseInserts: Scanner error: %v", err)
		return nil, err
	}
	if bar != nil {
		bar.SetTotal(bytesRead, true)
	}

	logger.Debug("ParseInserts: Processed %d INSERT statements, extracted %d rows across %d tables", acc.statementsProcessed, rowsExtracted, len(tableRows))
	return tableRows, nil
}

// ExpandInsert parses a single INSERT statement into multiple rows.
func (ip *InsertParser) ExpandInsert(insertStmt string, columnsMap map[string][]string) ([]*InsertRow, error) {
	// Normalize whitespace
	normalized := strings.Join(strings.Fields(insertStmt), " ")
	normalized = strings.TrimSpace(normalized)

	// Extract table name
	upperStmt := strings.ToUpper(normalized)
	insertIdx := strings.Index(upperStmt, "INSERT INTO")
	if insertIdx == -1 {
		return nil, fmt.Errorf("not an INSERT statement")
	}

	// Find table name after backtick
	restOfStmt := normalized[insertIdx+11:] // Skip "INSERT INTO"
	restOfStmt = strings.TrimSpace(restOfStmt)

	var table string
	if strings.HasPrefix(restOfStmt, "`") {
		endIdx := strings.Index(restOfStmt[1:], "`")
		if endIdx == -1 {
			return nil, fmt.Errorf("invalid table name")
		}
		table = restOfStmt[1 : endIdx+1]
		restOfStmt = restOfStmt[endIdx+2:]
	} else {
		// Table name without backticks
		parts := strings.FieldsFunc(restOfStmt, func(r rune) bool { return r == '(' || r == ' ' })
		if len(parts) == 0 {
			return nil, fmt.Errorf("invalid table name")
		}
		table = parts[0]
	}

	logger.Debug("ExpandInsert: Parsing INSERT for table: %s", table)

	// Extract columns and values
	var columns []string

	// Check if columns are explicit
	restOfStmt = strings.TrimSpace(restOfStmt)
	if strings.HasPrefix(restOfStmt, "(") {
		// Find matching closing paren for columns
		depth := 0
		var endIdx int
		for i, c := range restOfStmt {
			if c == '(' {
				depth++
			} else if c == ')' {
				depth--
				if depth == 0 {
					endIdx = i
					break
				}
			}
		}
		if endIdx == 0 {
			return nil, fmt.Errorf("invalid column list")
		}

		colStr := restOfStmt[1:endIdx]
		for _, col := range strings.Split(colStr, ",") {
			col = strings.TrimSpace(col)
			col = strings.Trim(col, "`'\"")
			if col != "" {
				columns = append(columns, col)
			}
		}

		restOfStmt = restOfStmt[endIdx+1:]
	}

	// Get values section
	restOfStmt = strings.TrimSpace(restOfStmt)
	valuesIdx := strings.ToUpper(restOfStmt)
	valuesPos := strings.Index(valuesIdx, "VALUES")
	if valuesPos == -1 {
		return nil, fmt.Errorf("no VALUES clause found")
	}

	valuesPart := restOfStmt[valuesPos+6:] // Skip "VALUES"
	valuesPart = strings.TrimSpace(valuesPart)

	// Remove trailing semicolon
	if strings.HasSuffix(valuesPart, ";") {
		valuesPart = valuesPart[:len(valuesPart)-1]
	}
	valuesPart = strings.TrimSpace(valuesPart)

	// If no explicit columns, use from schema
	if len(columns) == 0 {
		if columnsMap != nil && len(columnsMap[table]) > 0 {
			columns = columnsMap[table]
		}
	}

	if len(columns) == 0 {
		logger.Debug("ExpandInsert: No columns defined for table %s", table)
		return nil, fmt.Errorf("no columns defined for table %s", table)
	}

	// Split value groups
	groups := splitValueGroupsWithQuotes(valuesPart)
	logger.Debug("ExpandInsert: Table %s has %d value groups", table, len(groups))

	var rows []*InsertRow
	for _, group := range groups {
		group = strings.TrimSpace(group)
		if strings.HasPrefix(group, "(") && strings.HasSuffix(group, ")") {
			group = group[1 : len(group)-1]
		}

		values := parseValuesWithQuotes(group)
		if len(values) != len(columns) {
			logger.Debug("ExpandInsert: Skipping row with mismatched column count (expected %d, got %d)", len(columns), len(values))
			logger.Debug("ExpandInsert: Columns: %v", columns)
			logger.Debug("ExpandInsert: Values sample: %v", values[:min(5, len(values))])
			logger.Debug("ExpandInsert: Group snippet: %.200s...", group)
			continue
		}

		data := make(map[string]string)
		for i, col := range columns {
			data[col] = values[i]
		}

		rows = append(rows, &InsertRow{
			Table:   table,
			Columns: columns,
			Data:    data,
		})
	}

	logger.Debug("ExpandInsert: Table %s created %d rows", table, len(rows))
	return rows, nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// splitValueGroupsWithQuotes splits value groups while respecting quote boundaries
func splitValueGroupsWithQuotes(valuesPart string) []string {
	var groups []string
	var buf strings.Builder
	inSingleQuote := false
	inDoubleQuote := false
	escapeNext := false
	parenDepth := 0

	for _, c := range valuesPart {
		if escapeNext {
			buf.WriteRune(c)
			escapeNext = false
			continue
		}
		if c == '\\' {
			buf.WriteRune(c)
			escapeNext = true
			continue
		}
		if c == '\'' && !inDoubleQuote {
			buf.WriteRune(c)
			inSingleQuote = !inSingleQuote
			continue
		}
		if c == '"' && !inSingleQuote {
			buf.WriteRune(c)
			inDoubleQuote = !inDoubleQuote
			continue
		}
		if !inSingleQuote && !inDoubleQuote {
			if c == '(' {
				parenDepth++
				buf.WriteRune(c)
				continue
			}
			if c == ')' {
				parenDepth--
				buf.WriteRune(c)
				if parenDepth == 0 {
					groups = append(groups, buf.String())
					buf.Reset()
				}
				continue
			}
			if c == ',' && parenDepth == 0 {
				continue // Skip separator between groups
			}
		}
		buf.WriteRune(c)
	}

	if buf.Len() > 0 {
		remaining := strings.TrimSpace(buf.String())
		if remaining != "" {
			groups = append(groups, remaining)
		}
	}

	return groups
}

// parseValuesWithQuotes parses individual values while respecting quote boundaries
// Keeps quotes in values to match Java parser behavior for proper comparison
func parseValuesWithQuotes(valuesStr string) []string {
	var values []string
	var buf strings.Builder
	inSingleQuote := false
	inDoubleQuote := false
	escapeNext := false
	depth := 0 // Track parenthesis depth for nested structures

	for _, c := range valuesStr {
		if escapeNext {
			buf.WriteRune(c)
			escapeNext = false
			continue
		}
		if c == '\\' {
			buf.WriteRune(c)
			escapeNext = true
			continue
		}
		if c == '\'' && !inDoubleQuote && depth == 0 {
			buf.WriteRune(c) // Include the quote in the value (unlike previous version)
			inSingleQuote = !inSingleQuote
			continue
		}
		if c == '"' && !inSingleQuote && depth == 0 {
			buf.WriteRune(c) // Include the quote in the value
			inDoubleQuote = !inDoubleQuote
			continue
		}

		// Track parenthesis for nested structures
		if !inSingleQuote && !inDoubleQuote {
			if c == '(' {
				depth++
			} else if c == ')' {
				depth--
			}
		}

		if !inSingleQuote && !inDoubleQuote && depth == 0 && c == ',' {
			val := strings.TrimSpace(buf.String())
			if val == "NULL" {
				values = append(values, "NULL")
			} else {
				values = append(values, val)
			}
			buf.Reset()
		} else {
			buf.WriteRune(c)
		}
	}

	// Last value
	val := strings.TrimSpace(buf.String())
	if val == "NULL" {
		values = append(values, "NULL")
	} else {
		values = append(values, val)
	}

	return values
}

// ToJSON converts InsertRow to JSON for storage
func (ir *InsertRow) ToJSON() (string, error) {
	b, err := json.Marshal(ir)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// FromJSON parses JSON back into InsertRow
func FromJSON(jsonStr string) (*InsertRow, error) {
	var row InsertRow
	if err := json.Unmarshal([]byte(jsonStr), &row); err != nil {
		return nil, err
	}
	return &row, nil
}

File: rust/src/parser/insert.rs
// INSERT parser: streams INSERT statements and expands them into row objects.
// We intentionally keep parsing simple (no full SQL grammar) for speed.

use crate::logger;
use crate::parser::InsertRow;
use regex::Regex;
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};

pub struct InsertParser;

impl InsertParser {
    pub fn new() -> Self {
        Self
    }

    // Parse INSERT statements from a file, streaming rows to the caller.
    // If a progress bar is provided, we increment it by bytes read.
    pub fn parse_inserts_stream<F>(
        &self,
        filename: &str,
        columns_map: &HashMap<String, Vec<String>>,
        bar: Option<&indicatif::ProgressBar>,
        label: &str,
        mut on_row: F,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
    where
        F: FnMut(InsertRow),
    {
        logger::debug(&format!("ParseInsertsStream: Opening file {}", filename));
        let file = File::open(filename)?;
        let mut reader = BufReader::new(file);

        let mut acc = InsertAccumulator::new();
        let mut bytes_read: u64 = 0;
        let mut last_logged: u64 = 0;

        let mut line = String::new();
        while reader.read_line(&mut line)? > 0 {
            let line_len = line.len() as u64;
            bytes_read += line_len;
            if let Some(b) = bar {
                b.inc(line_len);
            } else if logger::is_debug() && bytes_read - last_logged > 100 * 1024 * 1024 {
                logger::debug(&format!(
                    "ParseInsertsStream: {} bytes read for {}",
                    bytes_read, label
                ));
                last_logged = bytes_read;
            }

            let statements = acc.process_line(&line);
            for stmt in statements {
                let rows = self.expand_insert(&stmt, columns_map);
                match rows {
                    Ok(rows) => {
                        for row in rows {
                            on_row(row);
                        }
                    }
                    Err(e) => {
                        logger::debug(&format!(
                            "ParseInsertsStream: failed to expand INSERT: {}",
                            e
                        ));
                    }
                }
            }

            line.clear();
        }

        // Flush any pending statement at EOF.
        for stmt in acc.finalize() {
            let rows = self.expand_insert(&stmt, columns_map);
            match rows {
                Ok(rows) => {
                    for row in rows {
                        on_row(row);
                    }
                }
                Err(e) => {
                    logger::debug(&format!(
                        "ParseInsertsStream: failed to expand INSERT: {}",
                        e
                    ));
                }
            }
        }

        if let Some(b) = bar {
            b.finish();
        }

        logger::debug(&format!(
            "ParseInsertsStream: processed {} statements",
            acc.statements_processed
        ));
        Ok(())
    }

    // Expand a single INSERT statement into one InsertRow per value group.
    pub fn expand_insert(
        &self,
        insert_stmt: &str,
        columns_map: &HashMap<String, Vec<String>>,
    ) -> Result<Vec<InsertRow>, String> {
        // Normalize whitespace for simpler parsing.
        let normalized = insert_stmt
            .split_whitespace()
            .collect::<Vec<_>>()
            .join(" ")
            .trim()
            .to_string();

        let upper = normalized.to_uppercase();
        let insert_idx = upper.find("INSERT INTO").ok_or("not INSERT")?;
        let mut rest = normalized[insert_idx + 11..].trim().to_string();

        // Parse table name.
        let table: String;
        if rest.starts_with('`') {
            let end = rest[1..]
                .find('`')
                .ok_or("invalid table name")?
                + 1;
            table = rest[1..end].to_string();
            rest = rest[end + 1..].to_string();
        } else {
            // Split by '(' or space.
            let mut parts = rest.split(|c: char| c == '(' || c.is_whitespace());
            table = parts.next().ok_or("invalid table name")?.to_string();
        }

        logger::debug(&format!("ExpandInsert: table {}", table));

        // Parse explicit columns if present.
        let mut columns: Vec<String> = Vec::new();
        let mut rest_trim = rest.trim().to_string();
        if rest_trim.starts_with('(') {
            let mut depth = 0i32;
            let mut end_idx = None;
            for (i, c) in rest_trim.chars().enumerate() {
                if c == '(' {
                    depth += 1;
                } else if c == ')' {
                    depth -= 1;
                    if depth == 0 {
                        end_idx = Some(i);
                        break;
                    }
                }
            }
            let end = end_idx.ok_or("invalid column list")?;
            let col_str = &rest_trim[1..end];
            for col in col_str.split(',') {
                let col = col.trim().trim_matches(['`', '\'', '"'].as_ref());
                if !col.is_empty() {
                    columns.push(col.to_string());
                }
            }
            rest_trim = rest_trim[end + 1..].to_string();
        }

        // Find VALUES keyword.
        let upper_rest = rest_trim.to_uppercase();
        let values_pos = upper_rest
            .find("VALUES")
            .ok_or("no VALUES clause")?;
        let mut values_part = rest_trim[values_pos + 6..].trim().to_string();
        if values_part.ends_with(';') {
            values_part.pop();
        }
        values_part = values_part.trim().to_string();

        // Use schema columns if INSERT has no explicit columns.
        if columns.is_empty() {
            if let Some(cols) = columns_map.get(&table) {
                columns = cols.clone();
            } else {
                // Case-insensitive fallback for table name mismatches.
                let table_lower = table.to_lowercase();
                for (key, cols) in columns_map.iter() {
                    if key.to_lowercase() == table_lower {
                        columns = cols.clone();
                        break;
                    }
                }
            }
        }
        if columns.is_empty() {
            return Err(format!("no columns defined for table {}", table));
        }

        let groups = split_value_groups_with_quotes(&values_part);
        logger::debug(&format!(
            "ExpandInsert: {} has {} value groups",
            table,
            groups.len()
        ));

        let mut rows = Vec::new();
        for mut group in groups {
            group = group.trim().to_string();
            if group.starts_with('(') && group.ends_with(')') {
                group = group[1..group.len() - 1].to_string();
            }
            let values = parse_values_with_quotes(&group);
            if values.len() != columns.len() {
                logger::debug(&format!(
                    "ExpandInsert: skip row (expected {}, got {})",
                    columns.len(),
                    values.len()
                ));
                continue;
            }
            let mut data: HashMap<String, String> = HashMap::new();
            for (i, col) in columns.iter().enumerate() {
                data.insert(col.clone(), values[i].clone());
            }
            rows.push(InsertRow {
                table: table.clone(),
                columns: columns.clone(),
                data,
            });
        }

        Ok(rows)
    }
}

// Tracks an INSERT statement across multiple lines.
struct InsertAccumulator {
    buffer: String,
    in_single_quote: bool,
    in_double_quote: bool,
    escape_next: bool,
    paren_depth: i32,
    in_insert: bool,
    statements_processed: usize,
}

impl InsertAccumulator {
    fn new() -> Self {
        Self {
            buffer: String::new(),
            in_single_quote: false,
            in_double_quote: false,
            escape_next: false,
            paren_depth: 0,
            in_insert: false,
            statements_processed: 0,
        }
    }

    // Process a single line and return any complete INSERT statements found.
    fn process_line(&mut self, line: &str) -> Vec<String> {
        let mut results = Vec::new();
        if !self.in_insert && line.to_uppercase().contains("INSERT INTO") {
            self.buffer.clear();
            self.buffer.push_str(line);
            self.in_insert = true;
            self.reset_state();
            self.process_line_content(line);
            if line.trim_end().ends_with(';') && self.paren_depth == 0 {
                results.push(self.buffer.clone());
                self.buffer.clear();
                self.in_insert = false;
                self.statements_processed += 1;
            }
            return results;
        }

        if self.in_insert {
            self.buffer.push_str(line);
            self.process_line_content(line);
            if line.trim_end().ends_with(';') && self.paren_depth == 0 {
                results.push(self.buffer.clone());
                self.buffer.clear();
                self.in_insert = false;
                self.statements_processed += 1;
            }
        }
        results
    }

    // Scan characters to keep track of quotes and parenthesis depth.
    fn process_line_content(&mut self, line: &str) {
        for c in line.chars() {
            if self.escape_next {
                self.escape_next = false;
                continue;
            }
            if c == '\\' {
                self.escape_next = true;
                continue;
            }
            if c == '\'' && !self.in_double_quote {
                self.in_single_quote = !self.in_single_quote;
            } else if c == '"' && !self.in_single_quote {
                self.in_double_quote = !self.in_double_quote;
            } else if !self.in_single_quote && !self.in_double_quote {
                if c == '(' {
                    self.paren_depth += 1;
                } else if c == ')' {
                    self.paren_depth -= 1;
                }
            }
        }
    }

    // Flush remaining buffer at EOF (best effort).
    fn finalize(&mut self) -> Vec<String> {
        let mut results = Vec::new();
        if self.in_insert && !self.buffer.is_empty() {
            results.push(self.buffer.clone());
            self.buffer.clear();
            self.statements_processed += 1;
        }
        results
    }

    fn reset_state(&mut self) {
        self.in_single_quote = false;
        self.in_double_quote = false;
        self.escape_next = false;
        self.paren_depth = 0;
    }
}

// Split value groups: (...),(...),... while respecting quotes.
fn split_value_groups_with_quotes(values_part: &str) -> Vec<String> {
    let mut groups = Vec::new();
    let mut buf = String::new();
    let mut in_single = false;
    let mut in_double = false;
    let mut escape = false;
    let mut paren_depth = 0i32;

    for c in values_part.chars() {
        if escape {
            buf.push(c);
            escape = false;
            continue;
        }
        if c == '\\' {
            buf.push(c);
            escape = true;
            continue;
        }
        if c == '\'' && !in_double {
            buf.push(c);
            in_single = !in_single;
            continue;
        }
        if c == '"' && !in_single {
            buf.push(c);
            in_double = !in_double;
            continue;
        }
        if !in_single && !in_double {
            if c == '(' {
                paren_depth += 1;
                buf.push(c);
                continue;
            }
            if c == ')' {
                paren_depth -= 1;
                buf.push(c);
                if paren_depth == 0 {
                    groups.push(buf.clone());
                    buf.clear();
                }
                continue;
            }
            if c == ',' && paren_depth == 0 {
                // Separator between groups; skip.
                continue;
            }
        }
        buf.push(c);
    }

    if !buf.trim().is_empty() {
        groups.push(buf.trim().to_string());
    }
    groups
}

// Parse comma-separated values within a single group, respecting quotes.
// We intentionally keep quotes in the value (matches Go/Java behavior).
fn parse_values_with_quotes(values_str: &str) -> Vec<String> {
    let mut values = Vec::new();
    let mut buf = String::new();
    let mut in_single = false;
    let mut in_double = false;
    let mut escape = false;
    let mut depth = 0i32;

    for c in values_str.chars() {
        if escape {
            buf.push(c);
            escape = false;
            continue;
        }
        if c == '\\' {
            buf.push(c);
            escape = true;
            continue;
        }
        if c == '\'' && !in_double && depth == 0 {
            buf.push(c);
            in_single = !in_single;
            continue;
        }
        if c == '"' && !in_single && depth == 0 {
            buf.push(c);
            in_double = !in_double;
            continue;
        }
        if !in_single && !in_double {
            if c == '(' {
                depth += 1;
            } else if c == ')' {
                depth -= 1;
            }
        }
        if !in_single && !in_double && depth == 0 && c == ',' {
            let val = buf.trim().to_string();
            values.push(if val == "NULL" { "NULL".to_string() } else { val });
            buf.clear();
        } else {
            buf.push(c);
        }
    }

    let val = buf.trim().to_string();
    values.push(if val == "NULL" { "NULL".to_string() } else { val });
    values
}

// Utility for testing: detect INSERT lines quickly.
#[allow(dead_code)]
fn is_insert_line(line: &str) -> bool {
    let re = Regex::new(r"(?i)INSERT\s+INTO").unwrap();
    re.is_match(line)
}
